src/main/java/bio/terra/service/iam/IamService.java

```diff
@@ -3,37 +3,47 @@
 import bio.terra.model.PolicyModel;
 import bio.terra.model.UserStatusInfo;
 import bio.terra.service.iam.exception.IamUnauthorizedException;
+import bio.terra.service.iam.exception.IamUnavailableException;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
 
 import java.util.List;
 import java.util.UUID;
 
-/**
- * This is the interface to IAM used in the main body of the repository code.
- * Right now, the only implementation of this service is SAM, but we expect
- * another implementation to be needed as part of the Framework work.
- */
-public interface IamService {
+@Component
+public class IamService {
+    private final IamProviderInterface iamProvider;
+
+    @Autowired
+    public IamService(IamProviderInterface iamProvider) {
+        this.iamProvider = iamProvider;
+    }
 
     /**
      * Is a user authorized to do an action on a resource.
      * @return true if authorized, false otherwise
      */
-    boolean isAuthorized(AuthenticatedUserRequest userReq,
+    public boolean isAuthorized(AuthenticatedUserRequest userReq,
                          IamResourceType iamResourceType,
                          String resourceId,
-                         IamAction action);
-
+                         IamAction action) {
+        try {
+            return iamProvider.isAuthorized(userReq, iamResourceType, resourceId, action);
+        } catch (InterruptedException ex) {
+            throw new IamUnavailableException("service unavailable");
```

Comment:
Can you explain your reasoning for converting an Interrupted -> IamUnavailable exception? I don't follow why we treat interrupts here differently than in other places throughout the code.

Reply:
For code that runs inside of flights, we want to let InterruptedException through so Stairway can catch it.

The iam code runs in both in flights and is called from the REST API. In the REST API controller, I could add InterruptedException to the method signature, since it derives from the swagger-codegen definition; that definition does not have the `throws`. So I needed to convert the exception to a RuntimeException for the REST API case.

I started to do that in the controller code, but it got pretty ugly. So I made a real IamService that handled the exception. In the REST API case, we use the service. In the Flight case, we use the provider.

I can add that as documentation at the top of IamService.

Reply:
That makes sense, yes.